src/app/icd/ICDManager.h

```diff
@@ -83,6 +80,18 @@ class ICDManager
      */
     static void OnTransitionToIdle(System::Layer * aLayer, void * appState);
 
+    /**
+     * @brief This function locks the ChipStack if the current thread didn't already do so.
+     *
+     *  Use case: Requests or notifications from the ICDNotifier can occur from a thread owning the lock already
```

Comment:
Generally I would let the caller have a `DoFoo` and `DoFooLocked` or `DoFooAsync` to specifify if we are whithin the chip stack or not. To get into the chip stack, I believe we generally schedule a work to be performed on the chip event loop.